shotover-proxy/src/frame/cassandra.rs

```diff
@@ -471,38 +503,499 @@ impl CassandraOperation {
+            }
+            Operand::Param(_) => true,
+            _ => false,
+        }
+    }
+
+    fn has_params_in_relation_elements(where_clause: &[RelationElement]) -> bool {
+        for relation_idx in where_clause {
+            if CQLStatement::has_params_in_operand(&relation_idx.value) {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Returns true if there are any parameters in the query
+    pub fn has_params(statement: &CassandraStatement) -> bool {
+        match statement {
+            CassandraStatement::Delete(delete) => {
+                if CQLStatement::has_params_in_relation_elements(&delete.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&delete.if_clause) {
+                    return true;
+                }
+            }
+            CassandraStatement::Insert(insert) => {
+                if let InsertValues::Values(operands) = &insert.values {
+                    for operand in operands {
+                        if let Operand::Param(_) = operand {
+                            return true;
+                        }
+                    }
+                }
+            }
+            CassandraStatement::Select(select) => {
+                return CQLStatement::has_params_in_relation_elements(&select.where_clause);
+            }
+            CassandraStatement::Update(update) => {
+                for assignment_element in &update.assignments {
+                    if let Operand::Param(_) = &assignment_element.value {
+                        return true;
+                    }
+                    if let Some(assignment_operator) = &assignment_element.operator {
+                        match assignment_operator {
+                            AssignmentOperator::Plus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                            AssignmentOperator::Minus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.if_clause) {
+                    return true;
+                }
+            }
+            _ => {}
+        }
+        false
+    }
+}
+
+impl Display for CQLStatement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.statement.fmt(f)
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct CQL {
+    pub statements: Vec<Box<CQLStatement>>,
```

Comment:
Replace the `Vec<Box<CQLStatement>>` with `Vec<CQLStatement>`
Theres no need for each individual CQLStatement to be given its own allocation on the heap

Reply:
Surprised clippy didn't catch this.

Reply:
I can think of some valid cases for it.
You might have values that are already boxed for example.
But that is not the case here.

Reply:
oh thanks, I'll give it a go myself and take a look at the clippy warnings